src/Features/Core/Portable/MetadataAsSource/DecompilationMetadataAsSourceFileProvider.cs

```diff
@@ -292,32 +293,30 @@ public bool ShouldCollapseOnOpen(MetadataAsSourceWorkspace workspace, string fil
         return false;
     }
 
-    public bool TryAddDocumentToWorkspace(MetadataAsSourceWorkspace workspace, string filePath, SourceTextContainer sourceTextContainer)
+    public bool TryAddDocumentToWorkspace(MetadataAsSourceWorkspace workspace, string filePath, SourceTextContainer sourceTextContainer, [NotNullWhen(true)] out DocumentId? documentId)
     {
-        AssertIsMainThread(workspace);
```

Comment:
Wasn't quite sure what to do with all these asserts. The callers of these in the VS layer assert the main thread, but we can't assert the main thread from LSP.  So I removed the asserts.  Not sure if there is a better way.

Reply:
Just put locks around all the methods, if possible? (I haven't looked at the code in detail so this suggestion could be terrible.)

Reply:
yeah, i agree with jason.  fine to be synchronized.